rustc_codegen_spirv/src/linker/structurizer.rs

```diff
@@ -1,17 +1,169 @@
 // This pass inserts merge instructions for structured control flow with the assumption the spir-v is reducible.
 
+// TODO: Could i simplify break detection by just checking, hey does the start branch branch to a merge block?
+// TODO: Verify we are never splitting a block that is queued for structurization.
+// TODO: are there any cases where I need to retarget branches or conditional branches when splitting a block?
+
 use super::id;
 use super::simple_passes::outgoing_edges;
 use rspirv::spirv::{Op, Word};
 use rspirv::{
-    dr::{Block, Function, Instruction, ModuleHeader, Operand},
+    dr::{Block, Instruction, Module, ModuleHeader, Operand},
     spirv::SelectionControl,
 };
 use rustc_session::Session;
 use std::collections::VecDeque;
 
-pub fn structurize(sess: Option<&Session>, header: &mut ModuleHeader, func: &mut Function) {
-    insert_selection_merge_on_conditional_branch(sess, header, &mut func.blocks);
+pub struct LoopInfo {
+    merge_id: Word,
+    continue_id: Word,
+    header_id: Word,
+}
+pub struct ControlFlowInfo {
+    loops: Vec<LoopInfo>,
+    if_merge_ids: Vec<Word>,
+    switch_merge_ids: Vec<Word>,
```

Comment:
Should these be HashSets? Quick scan looks like there's no order-dependent stuff going on, and hashset would significantly improve the runtime of the `id_is_loops_merge` functions and friends.

Reply:
It may be faster to do something like `IndexVec<Word, Option<ControlFlowKind>>` assuming that any `Word` can only be one of loop merge, loop continue, loop header, if merge, switch merge or none of the former.